homeassistant/components/homekit/__init__.py

```diff
@@ -3,154 +3,197 @@
+from .util import (
+    validate_entity_config, show_setup_message)
 
 TYPES = Registry()
 _LOGGER = logging.getLogger(__name__)
 
-_RE_VALID_PINCODE = r"^(\d{3}-\d{2}-\d{3})$"
-
-DOMAIN = 'homekit'
-REQUIREMENTS = ['HAP-python==1.1.7']
-
-BRIDGE_NAME = 'Home Assistant'
-CONF_PIN_CODE = 'pincode'
-
-HOMEKIT_FILE = '.homekit.state'
-
-
-def valid_pin(value):
-    """Validate pin code value."""
-    match = re.match(_RE_VALID_PINCODE, str(value).strip())
-    if not match:
-        raise vol.Invalid("Pin must be in the format: '123-45-678'")
-    return match.group(0)
+REQUIREMENTS = ['HAP-python==1.1.7', 'pypng==0.0.18']
 
 
 CONFIG_SCHEMA = vol.Schema({
     DOMAIN: vol.All({
-        vol.Optional(CONF_PORT, default=51826): vol.Coerce(int),
-        vol.Optional(CONF_PIN_CODE, default='123-45-678'): valid_pin,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): vol.Coerce(int),
+        vol.Optional(CONF_AUTO_START, default=DEFAULT_AUTO_START): cv.boolean,
+        vol.Optional(CONF_FILTER, default={}): FILTER_SCHEMA,
+        vol.Optional(CONF_ENTITY_CONFIG, default={}): validate_entity_config,
     })
 }, extra=vol.ALLOW_EXTRA)
 
 
-@asyncio.coroutine
-def async_setup(hass, config):
+async def async_setup(hass, config):
     """Setup the HomeKit component."""
-    _LOGGER.debug("Begin setup HomeKit")
+    _LOGGER.debug('Begin setup HomeKit')
 
     conf = config[DOMAIN]
-    port = conf.get(CONF_PORT)
-    pin = str.encode(conf.get(CONF_PIN_CODE))
+    port = conf[CONF_PORT]
+    auto_start = conf[CONF_AUTO_START]
+    entity_filter = conf[CONF_FILTER]
+    entity_config = conf[CONF_ENTITY_CONFIG]
 
-    homekit = HomeKit(hass, port)
-    homekit.setup_bridge(pin)
+    homekit = HomeKit(hass, port, entity_filter, entity_config)
+    homekit.setup()
 
-    hass.bus.async_listen_once(
-        EVENT_HOMEASSISTANT_START, homekit.start_driver)
-    return True
+    if auto_start:
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, homekit.start)
+        return True
+
+    def handle_homekit_service_start(service):
+        """Handle start HomeKit service call."""
+        if homekit.started:
+            _LOGGER.warning('HomeKit is already running')
+            return
+        homekit.start()
 
+    hass.services.async_register(DOMAIN, SERVICE_HOMEKIT_START,
+                                 handle_homekit_service_start)
 
-def import_types():
-    """Import all types from files in the HomeKit directory."""
-    _LOGGER.debug("Import type files.")
-    # pylint: disable=unused-variable
-    from . import (  # noqa F401
-        covers, security_systems, sensors, switches, thermostats)
+    return True
 
 
-def get_accessory(hass, state):
+def get_accessory(hass, state, aid, config):
     """Take state and return an accessory object if supported."""
+    _LOGGER.debug('%s: <aid=%d config=%s>')
+    if not aid:
+        _LOGGER.warning('The entitiy "%s" is not supported, since it '
```

Comment:
I'm surprised the warning is in here, shouldn't generation of aid also be in here then ? 

Also, under what circumstances do we generate an invalid aid?

Reply:
`zlib.adler32` can generate values in the range of `[0, 2**32-1]`. 0 is forbidden and an aid of 1 has the bridge. I wanted to outsource the generation to an own method `generate_aid`. It is called from `add_bridge_accessory` right before `get_accessory`.